### PR TITLE
Change predicate syntax to allow nested implications

### PR DESCRIPTION
**Pull request by dario23**
Constructs like Prusti's implication `==>` that aren't valid Rust were
not possible in previous style `#[predicate]` functions, because rustc
parsed the functions already for e.g. name resolution, before we were
able to rewrite them. With the new function-like proc-macro aproach, we
avoid that problem.

### CODE REVIEW

File: prusti-tests/tests/parse/ui/no-attr-params.rs [hunk 1]
```diff
@@ -13,7 +13,9 @@ fn pure_err() -> bool {
 #[trusted(dont write things here)]
 fn trusted_err() {}
 
-#[predicate(dont put anything here)]
-fn pred_err() {}
+predicate! {
+    #[dont_put_anything_here]
+    fn pred_err() {}
+}
```
Comment (Aurel300) [inline prusti-tests/tests/parse/ui/no-attr-params.rs]: I think this test can just be deleted now.

File: prusti-specs/src/lib.rs [hunk 2]
```diff
@@ -480,64 +480,55 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
     }
 }
 
-pub fn predicate(attr: TokenStream, tokens: TokenStream) -> TokenStream {
-    if !attr.is_empty() {
-        return syn::Error::new(
-            attr.span(),
-            "the `#[predicate]` attribute does not take parameters"
-        ).to_compile_error();
+// for syn::parse the predicate thingy
```
Comment (Aurel300) [inline prusti-specs/src/lib.rs]:
```suggestion
```

File: prusti-tests/tests/parse/ui/no-attr-params.rs (hunk 1 above)
Comment (Aurel300) [reply prusti-tests/tests/parse/ui/no-attr-params.rs]: Sorry, I should have been clearer – I meant the highlighted lines, the remainder of the file is still good.

Comment (dario23) [reply prusti-tests/tests/parse/ui/no-attr-params.rs]: ah, fixed.

**Comment by dario23** [comment]
test that fails is `[run-pass] pass/quick/enums.rs`, which i haven't changed anything about. `git bisect` thinks it's the commit merging the new viper-tools handling (#553). is there a significantly different viper/z3 version in that?

**Pull request merged**